lxd/device/nic_macvlan.go

```diff
@@ -0,0 +1,153 @@
+package device
+
+import (
+	"fmt"
+
+	"github.com/lxc/lxd/lxd/device/config"
+	"github.com/lxc/lxd/lxd/instance"
+	"github.com/lxc/lxd/shared"
+)
+
+type nicMACVLAN struct {
+	deviceCommon
+}
+
+// validateConfig checks the supplied config for correctness.
+func (d *nicMACVLAN) validateConfig() error {
+	if d.instance.Type() != instance.TypeContainer {
+		return ErrUnsupportedDevType
+	}
+
+	rules := map[string]func(value string) error{
```

Comment:
Still feels like we could have gone one step further for nics as the validation of the individual keys is likely identical across nictypes (with a few minor exceptions) so we could have centralized this so if we change the validator for a key (say `vlan` which would be an integer from 0 up to 4094) we don't need to go and update all of them.

I suspect this could be implement by a nic-specific function which is fed the list of required keys and list of valid keys and would return the rule map.

It would internally have the mapping from key name to the usual validator. If we need a nictype-specific validator, we can then easily override that one entry from the returned map before feeding it to ValidateDevice.

Reply:
@stgraber good idea, have added that now.